### PR TITLE
Add CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by michaelpj**
@mchakravarty has had a look at this content, but I'm keen to hear thoughts from others. This is largely based on my experience at other companies, but we can use whatever process we want!

### CODE REVIEW

**Comment by michaelpj** [comment]
Hm, it would be nice if the CI had `-Werror` set too, which I believe it currently doesn't.

**Comment by michaelpj** [comment]
I added a note about compiler warnings.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+- Try to review PRs where your review is requested within a few days. This should be nearly-top-priority work.
+- If you don’t understand something then ask for an explanation.
+    - For the author: this explanation should ideally be added as a comment - you’re going to write it anyway, and future readers are likely to be just as confused as the reviewer.
+
+# Continuous integration
+
+We have two pieces of CI at the moment: some tests are run using Nix on Hydra, and some are run on Buildkite.
+- All the projects will be built, as well as the tests in `default.nix`.
+- Pull requests cannot be merged without the CI going green.
+    - Because the CI is not necessarily run on the merge commit that is created when the PR is merged, it is possible that merging a green PR can result in the CI being broken on master. This shouldn't happen frequently, but be aware that it's possible.
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply CONTRIBUTING.md]: We used to do that! But it's very annoying that merging a pull request invalidates every single other one and requires manual action to update them.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+## Code style
+
+We use `stylish-haskell` and `hlint`, and enable a large number of GHC warnings.
+- These are run by the CI, so if you don’t use them your PR will not go green. To avoid annoyance, set up your editor to run them automatically.
+- It’s fine to be aggressive about disabling `hlint` rules that don’t provide value, but check with the team before doing this. Err on the side of disabling them globally rather than for specific files, so we stay consistent.
+- Try and make sure that you do not have any compiler warnings - this is not currently checked by the CI, but may be in future.
+
+# Issues
+
+We track our issues on the IOHK YouTrack instance: https://iohk.myjetbrains.com/youtrack/issues/CGP
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply CONTRIBUTING.md]: Yes, that seems reasonable. I'll clarify that it depends on whether you're inside IOHK or not.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+# Submitting changes
+
+## Pull requests
+
+All code changes go through pull requests.
+- Make your PR from the main repository if possible, this is necessary for the Buildkite CI to trust you.
+- PRs exist to be reviewed - design them with a reader in mind!
+    - Write a helpful PR description that explains what’s in the PR and why, and draws attention to anything of particular note.
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply CONTRIBUTING.md]: Good idea.

**Comment by effectfully** [comment]
It's probably worth adding a note about when IOHK developers should open YouTrack issues. Whenever you fix anything (please no)? Whenever you fix anything big? Whenever you're going to fix something, but not right now? Whenever you think there is something wrong with someone else's code? Other options?

**Comment by michaelpj** [comment]
Hm, the linked style guide is quite aggressive. I think we should leave that out for now, we can always add it if we decide we want to follow it.

I agree it would be nice to have something about when to file YouTrack issues, but perhaps @mchakravarty could add that as a follow up.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+`pkgs/default.nix` contains a generated package set with all the dependencies for this project.
+
+You should regenerate this if you change any dependencies in cabal files. To do this, use the environment defined in `shell.nix`, and run `pkgs/generate.sh`.
+
+## Code style
+
+We use `stylish-haskell` and `hlint`, and enable a large number of GHC warnings.
+- These are run by the CI, so if you don’t use them your PR will not go green. To avoid annoyance, set up your editor to run them automatically.
+- It’s fine to be aggressive about disabling `hlint` rules that don’t provide value, but check with the team before doing this. Err on the side of disabling them globally rather than for specific files, so we stay consistent.
+- Try and make sure that you do not have any compiler warnings - this is not currently checked by the CI, but may be in future.
```
Comment (j-mueller) [inline CONTRIBUTING.md]: Perhaps this should include a line on cabal files (`default-extensions`, `other-extensions`, `ghc-options`, and whether or not `-any` should be added to dependencies without bounds)

**Comment by michaelpj** [comment]
Can I ask that people not make more suggestions for additions? We can always add more things later, I'd like to make sure we agree on at least what's in here, and then go on from there.

**Comment by michaelpj** [comment]
I added a note about PRs from forks if you don't have write access.

Nobody's objected to what's actually in here so far, so I'll merge in ~a day and people can make more PRs for any additions.

**Comment by Nikolaj-K** [comment]
@michaelpj You had referenced an issue I opened here a month ago and so I was looking back at it. In semi-related news, I just saw the nice summary of PR suggestions. I'm looking for this kind of recommendations for my own projects as well, did you take it from somewhere or where can I find such thought through recommendations? Thanks.

**Comment by michaelpj** [comment]
No, this is just based on my own experience. You're welcome to lift things from it, though.

**Comment by Nikolaj-K** [comment]
Ah great. 

Another inquiry... I personally don't think that a particular consensus model that's created now will be the relevant or used one in 5 years (e.g. if a big company with a long existing product/usecase puts their own resources into creating a network, it will leave to current protocols behind), but a language and VM specifications could. From that perspective:
Do you have any links to material how to create a VM that interprets Plutus and e.g. which constracts it will be most suited to express?

**Comment by michaelpj** [comment]
You can find the Plutus Core specification in this repository, which will give you some of that information. But it's quite a lot of flux - I'd check back in a few months!

**Comment by Nikolaj-K** [comment]
Thanks again, I'll be monitoring.

**Pull request merged**